src/main/java/org/jabref/logic/pdf/search/indexing/DocumentReader.java

```diff
@@ -0,0 +1,138 @@
+ */
+public final class DocumentReader {
+    private static final Log LOGGER = LogFactory.getLog(DocumentReader.class);
+
+    private final BibEntry entry;
+    private final FilePreferences filePreferences;
+
+    /**
+     * Creates a new DocumentReader using a BibEntry.
+     *
+     * @param bibEntry Must not be null and must have at least one LinkedFile.
+     */
+    public DocumentReader(BibEntry bibEntry, FilePreferences filePreferences) {
+        this.filePreferences = filePreferences;
+        if (bibEntry.getFiles().isEmpty()) {
+            throw new IllegalStateException("There are no linked PDF files to this BibEntry!");
+        }
+
+        this.entry = bibEntry;
+    }
+
+    /**
+     * Reads a LinkedFile of a BibEntry and converts it into a Lucene Document which is then returned.
+     *
+     * @return An Optional of a Lucene Document with the (meta)data. Can be empty if there is a problem reading the LinkedFile.
+     */
+    public Optional<Document> readLinkedPdf(BibDatabaseContext databaseContext, LinkedFile pdf) {
+        Optional<Path> pdfPath = pdf.findIn(databaseContext, filePreferences);
+        if (pdfPath.isPresent()) {
+            try {
+                return Optional.of(readPdfContents(pdf, pdfPath.get()));
+            } catch (IOException e) {
+                LOGGER.info("Could not read pdf file: " + pdf.getLink() + "!", e);
+            }
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Reads each LinkedFile of a BibEntry and converts them into Lucene Documents which are then returned.
+     *
+     * @return A List of Documents with the (meta)data. Can be empty if there is a problem reading the LinkedFile.
+     */
+    public List<Document> readLinkedPdfs(BibDatabaseContext databaseContext) {
+        return entry.getFiles().stream()
+                    .map((pdf) -> readLinkedPdf(databaseContext, pdf))
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .collect(Collectors.toList());
+    }
+
+    private Document readPdfContents(LinkedFile pdf, Path resolvedPdfPath) throws IOException {
+        try (PDDocument pdfDocument = PDDocument.load(resolvedPdfPath.toFile())) {
+            Document newDocument = new Document();
+            addIdentifiers(newDocument, pdf.getLink());
+            addContentIfNotEmpty(pdfDocument, newDocument);
+            addMetaData(pdfDocument, newDocument, resolvedPdfPath);
+            return newDocument;
+        }
+    }
+
+    private void addMetaData(PDDocument pdfDocument, Document newDocument, Path resolvedPdfPath) {
+        try {
+            BasicFileAttributes attributes = Files.readAttributes(resolvedPdfPath, BasicFileAttributes.class);
+            addStringField(newDocument, MODIFIED, String.valueOf(attributes.lastModifiedTime().to(TimeUnit.SECONDS)));
+        } catch (IOException e) {
+            LOGGER.error("Could not read timestamp for " + resolvedPdfPath, e);
```

Comment:
I noticed that some classes use apache commons logging (outside of this PR only CleanupUrlFormatter) while most use slf4j. Was there a migration to slf4j from when this PR was originally started?
The thing is, what you are suggesting only works in slf4j, so if this is the case I would migrate the classes in this PR to slf4j.

Reply:
I noticed that in code-howtos.md, it says to use slf4j, but the given code is for apache (LogFactory.getLog instead of LoggerFactory.getLogger). Shall I fix that in a separate PR?

Reply:
I don't know when exactly the migration was, but you should use the slf4j. Slf4j is a facade for the underlying logging implementation (log4j in this case)

Reply:
This refs https://github.com/JabRef/jabref/issues/5475#issuecomment-858083303.

Reply:
I removed references to apache commons logging in #7907